lib/src/internal/http/http_bucket.dart

```diff
@@ -9,86 +9,67 @@ import 'package:nyxx/src/internal/http/http_handler.dart';
 import 'package:nyxx/src/internal/http/http_request.dart';
 
 class HttpBucket {
-  // Rate limits
-  int _remaining = 10;
-  DateTime? _resetAt;
-  double? _resetAfter;
-  RestEventController get _events => _httpHandler.client.eventsRest as RestEventController;
+  static const String xRateLimitBucket = "x-ratelimit-bucket";
+  static const String xRateLimitLimit = "x-ratelimit-limit";
+  static const String xRateLimitRemaining = "x-ratelimit-remaining";
+  static const String xRateLimitReset = "x-ratelimit-reset";
+  static const String xRateLimitResetAfter = "x-ratelimit-reset-after";
 
-  // Bucket ID
-  late final String id;
+  int _limit;
+  int _remaining;
+  DateTime _reset;
+  Duration _resetAfter;
+  final String _bucketId;
 
-  // Reference to http handler
-  final HttpHandler _httpHandler;
+  final List<HttpRequest> _inFlightRequests = [];
 
-  /// Creates an instance of [HttpBucket]
-  HttpBucket(this.id, this._httpHandler);
+  int get remaining => _remaining - _inFlightRequests.length;
 
-  Future<http.StreamedResponse> execute(HttpRequest request) async {
-    _httpHandler.logger.fine(
-        "Executing request: [${request.uri.toString()}]; Bucket ID: [$id]; Reset at: [$_resetAt]; Remaining: [$_remaining]; Reset after: [$_resetAfter]; Body: [${request is BasicRequest && request.body != null ? request.body : 'EMPTY'}]");
+  DateTime get reset => _reset;
 
-    // Get actual time and check if request can be executed based on data that bucket already have
-    // and wait if rate limit could be possibly hit
-    final now = DateTime.now();
-    if ((_resetAt != null && _resetAt!.isAfter(now)) && _remaining < 2) {
-      final waitTime = _resetAt!.millisecondsSinceEpoch - now.millisecondsSinceEpoch;
+  Duration get resetAfter => _resetAfter;
 
-      if (waitTime > 0) {
-        _events.onRateLimitedController.add(RatelimitEvent(request, true));
-        _httpHandler.logger.warning("Rate limited internally on endpoint: ${request.uri}. Trying to send request again in $waitTime ms...");
+  String get id => _bucketId;
 
-        return Future.delayed(Duration(milliseconds: waitTime), () => execute(request));
-      }
-    }
+  HttpBucket(this._limit, this._remaining, this._reset, this._resetAfter, this._bucketId);
+
+  static HttpBucket? fromResponseSafe(http.StreamedResponse response) {
+    int? limit = getLimitFromHeaders(response.headers);
+    int? remaining = getRemainingFromHeaders(response.headers);
+    DateTime? reset = getResetFromHeaders(response.headers);
+    Duration? resetAfter = getResetAfterFromHeaders(response.headers);
+    String? bucketId = getBucketIdFromHeaders(response.headers);
+    return limit == null || remaining == null || reset == null || resetAfter == null || bucketId == null ? null : HttpBucket(limit, remaining, reset, resetAfter, bucketId);
```

Comment:
that one liner is not readable. Can you maybe do proper if statement here?

Reply:
+1 This should not be one line.